amazon_cloud_code_generator/data/plugins/module_utils/core.py

```diff
@@ -461,31 +462,43 @@ def update_resource(
                         e, msg="Failed to cancel resource request"
                     )
 
-            self.wait_for_in_progress_requests(in_progress_requests, identifier)
+        patch = get_patch(self.module, params, properties)
+        obj, error = json_patch(properties, patch)
+        if error:
+            self.module.fail_json(**error)
+        match, diffs = diff_dicts(properties, obj)
 
-            try:
-                response = self.client.update_resource(
-                    TypeName=type_name,
-                    Identifier=identifier,
-                    PatchDocument=str(patch),
-                )
-                obj = json.loads(response["ProgressEvent"]["ResourceModel"])
-            except self.client.exceptions.ResourceNotFoundException:
-                # If there is an IN PROGRESS delete opration on the resource and
-                # the resource is deleted before (so it can't be updated)
-                return results
-            except (
-                botocore.exceptions.BotoCoreError,
-                botocore.exceptions.ClientError,
-            ) as e:
-                self.module.fail_json_aws(e, msg="Failed to update resource")
+        if not self.module.check_mode:
+            # To handle idempotency when purge_* params are False (where the patch is always generated with strategy='replace')
+            # call self.client.update_resource() only when there's a difference
+            if diffs:
+                # Wait for IN PROGRESS or PENDING resource requests to avoid concurrency exceptions
+                self.wait_for_in_progress_requests(in_progress_requests, identifier)
+                try:
+                    response = self.client.update_resource(
+                        TypeName=type_name,
+                        Identifier=identifier,
+                        PatchDocument=str(patch),
+                    )
+                except (
+                    botocore.exceptions.BotoCoreError,
+                    botocore.exceptions.ClientError,
+                ) as e:
+                    self.module.fail_json_aws(e, msg="Failed to update resource")
 
-        if not self.module.check_mode and self.module.params.get("wait"):
-            self.wait_until_resource_request_success(
-                response["ProgressEvent"]["RequestToken"]
-            )
+                # Ensure the request is at least IN_PROGRESS to return updated information
+                # Tag updates hangs on PENDING sometimes and updates are not reflected on the resource at this stage
+                self.ensure_request_status(response)
```

Comment:
Why wouldn't we put this line within the `if self.module.params.get("wait")` block two lines below?

Reply:
This is needed because when there's a tag update the resource request hangs on PENDING sometimes and the result retuned is not updated and the idempotency is broken. This helps when the user does not use wait=true. We should always wait to have a resource request IN PROGRESS at least, then, it's up to the user if waiting till the resource request reaches SUCCESS with wait=true.